semantics-3.0/Semantics.hs

```diff
@@ -236,206 +227,219 @@ data ReduceWarning = ReduceNoWarning
+
+    If obs cont1 cont2 -> let
+        cont = if evalObservation env state obs then cont1 else cont2
+        in Reduced ReduceNoWarning Nothing state cont
+
+    When _ timeout cont -> let
+        startSlot = ivFrom (slotInterval env)
+        endSlot   = ivTo (slotInterval env)
         -- if timeout in future – do not reduce
-        | endSlot < timeout -> NotReduced
+        in if endSlot < timeout then NotReduced
         -- if timeout in the past – reduce to timeout continuation
-        | timeout <= startSlot -> Reduced ReduceNoWarning ReduceNoEffect state cont
+        else if timeout <= startSlot then Reduced ReduceNoWarning Nothing state cont
         -- if timeout in the slot range – issue an ambiguity error
-        | otherwise -> ReduceError ReduceAmbiguousSlotInterval
-      where
-        startSlot = fst $ slotInterval env
-        endSlot   = snd $ slotInterval env
-    Let valId val cont -> Reduced warn ReduceNoEffect newState cont
-      where
-        evVal = evalValue env state val
+        else AmbiguousSlotIntervalError
+
+    Let valId val cont -> let
+        evaluatedValue = evalValue env state val
         boundVals = boundValues state
-        newState = state { boundValues = Map.insert valId evVal boundVals }
+        newState = state { boundValues = Map.insert valId evaluatedValue boundVals }
         warn = case Map.lookup valId boundVals of
-              Just oldVal -> ReduceShadowing valId oldVal evVal
+              Just oldVal -> ReduceShadowing valId oldVal evaluatedValue
               Nothing -> ReduceNoWarning
+        in Reduced warn Nothing newState cont
 
 
--- REDUCE ALL
-
-data ReduceAllResult = ReducedAll [ReduceWarning] [ReduceEffect] State Contract
-                     | ReduceAllError ReduceError
-  deriving (Eq,Ord,Show,Read)
+data ReduceAllResult = ReduceAllSuccess [ReduceWarning] [Payment] State Contract
+                     | ReduceAllAmbiguousSlotIntervalError
+  deriving (Eq,Ord,Show)
 
--- Reduce until it cannot be reduced more
-reduceAllAux
-  :: Environment -> State -> Contract -> [ReduceWarning] -> [ReduceEffect] -> ReduceAllResult
-reduceAllAux env state contract warnings effects = case reduce env state contract of
-    Reduced warning effect newState cont -> let
-        newWarnings = if warning == ReduceNoWarning then warnings else warning : warnings
-        newEffects  = if effect  == ReduceNoEffect  then effects  else effect : effects
-        in  reduceAllAux env newState cont newWarnings newEffects
-    ReduceError err -> ReduceAllError err
-    NotReduced -> ReducedAll (reverse warnings) (reverse effects) state contract
+-- | Reduce a contract until it cannot be reduced more
+reduceContractUntilQuiescent :: Environment -> State -> Contract -> ReduceAllResult
+reduceContractUntilQuiescent env state contract = let
+    reduceAll
+      :: Environment -> State -> Contract -> [ReduceWarning] -> [Payment] -> ReduceAllResult
+    reduceAll env state contract warnings effects =
```

Comment:
If reduceAll is a bad name, it will still be a bad name for a sub-function. We could just call it "loop" or something like that if it is inlined.

In Isabelle it is still probably better to keep it as a separate function, but we do not need to get restricted by that in the Haskell semantics.

Reply:
Also, why do we use the word "effect" rather than "payment"? I think we discussed it but I can't remember. I bring this up again because here we are using the type `[Payment]` with the term name `effects`.